src/runtime/instance.rs

```diff
@@ -0,0 +1,183 @@
+                return Err("incorrect signature".to_string());
+            }
+        }
+
+        // the vmctx will be located at the same place on the stack the entire time that this
+        // wasm function is running.
+        let mut vmctx = vm::Ctx::new(
+            &mut self.backing,
+            &mut self.import_backing,
+            &self.sig_registry,
+        );
+        let vmctx_ptr = &mut vmctx as *mut vm::Ctx;
+
+        let libffi_args: Vec<_> = args
+            .iter()
+            .map(|val| match val {
+                Value::I32(ref x) => libffi_arg(x),
+                Value::I64(ref x) => libffi_arg(x),
+                Value::F32(ref x) => libffi_arg(x),
+                Value::F64(ref x) => libffi_arg(x),
+            })
+            .chain(iter::once(libffi_arg(&vmctx_ptr)))
+            .collect();
+
+        let func_ptr = CodePtr::from_ptr(
+            self.module
+                .func_resolver
+                .get(&*self.module, func_index)
+                .expect("broken invariant, func resolver not synced with module.exports")
+                .cast()
+                .as_ptr(),
+        );
+
+        call_protected(|| {
+            self.module.signatures[sig_index]
+                .returns
+                .first()
+                .map(|ty| match ty {
+                    Type::I32 => Value::I32(unsafe { libffi_call(func_ptr, &libffi_args) }),
+                    Type::I64 => Value::I64(unsafe { libffi_call(func_ptr, &libffi_args) }),
+                    Type::F32 => Value::F32(unsafe { libffi_call(func_ptr, &libffi_args) }),
+                    Type::F64 => Value::F64(unsafe { libffi_call(func_ptr, &libffi_args) }),
+                })
+                .or_else(|| {
+                    // call with no returns
+                    unsafe {
+                        libffi_call::<()>(func_ptr, &libffi_args);
+                    }
+                    None
+                })
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Import {
+    Func(*const vm::Func, FuncSig),
+    Table(Arc<TableBacking>, Table),
+    Memory(Arc<LinearMemory>, Memory),
+    Global(Value),
+}
+
+// TODO Remove again
+pub struct Imports {
+    map: HashMap<String, HashMap<String, Import>>,
+}
+
+// TODO Remove again
+impl Imports {
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+        }
+    }
+
+    pub fn add(&mut self, module: String, name: String, import: Import) {
+        self.map
+            .entry(module)
+            .or_insert(HashMap::new())
+            .insert(name, import);
+    }
+
+    pub fn get(&self, module: &str, name: &str) -> Option<&Import> {
+        self.map.get(module).and_then(|m| m.get(name))
+    }
+}
+
+impl ImportResolver for Imports {
+    fn get(&self, module: &str, name: &str) -> Option<Import> {
+        unimplemented!("ImportResolver for Imports")
+    }
+}
+
+pub trait ImportResolver {
+    fn get(&self, module: &str, name: &str) -> Option<Import>;
```

Comment:
I noticed this went from `&Import` to `Import` should this be `&Import`?

Reply:
It could be either really, would you rather it be a ref?